### ISSUE TITLE
[BUG][CV] Check Similar image leakage threshold on several datasets (fails on MNIST)

### ISSUE THREAD

**Issue by noamzbr** [deepchecks/deepchecks#1243]
**Describe the bug**

When running Similar image leakage on MNIST, many many samples are found to be similar, probably because they are (a 8 written by two different people can conceivably get a very similar hash). This raised the question of is the threshold general.

**Desired Solution**
Not clear, but should probably involve having the threshold be calibrated according to the internal variance of hashes within the same class. Should be tested on several more datasets and see what threshold is good for them. If same threshold is good for all except MNIST

**To Preproduce**
See the test in https://github.com/deepchecks/deepchecks/pull/1241

**Comment by nirhutnik** [comment]
Ok, conclusions:

1. The check is very inefficient when there are many similarities and we should fix that. I added a small fix but that's not a good enough solution.
2. I searched for datasets, and specifically for ones that are "problematic" like mnist. I found this pistachio dataset - https://www.kaggle.com/datasets/muratkokludataset/pistachio-image-dataset/code . It's pictures of pistachios on a black background.
3. As can be seen in this image, many of the images are hashed to the SAME hash, as most of the information in the image is similar. 
![image](https://user-images.githubusercontent.com/92314933/164266373-9a9e5265-1622-4486-aaa8-3f1398578bdb.png)
4. When changing hash size, this is mostly fixed if using a reasonable similarity_threshold for this dataset:
![image](https://user-images.githubusercontent.com/92314933/164266746-e9bd7160-2db8-4c92-b165-ea526054703d.png)
5. However, this is still problematic - this kind of dataset, which is almost duochromatic, with a single object in its midst, is the exact Achilles' heel of this method: change is in hues that are either absolutely above or below the average, and images have very similar edges.
6. I think we should have a more thorough discussion on this. I'm scheduling one for Sunday. My directions are: some way to analyze the recommended hash size and similarity score, or using a different method for duochromatic datasets. But this requires more thinking.

**Comment by nirhutnik** [comment]
Code used:
(images were joined to a folder names all_images, and labels weren't really implements. Also, used my own path. Code wasn't meant to be general, just for my usage, so it's not the best :) )
```
import numpy as np
from deepchecks.vision.checks import SimilarImageLeakage
import cv2
from PIL import Image
from pathlib import Path
import random
from torch.utils.data import DataLoader
from torchvision.datasets import VisionDataset
from torch import tensor
from deepchecks.vision import ClassificationData

random.seed(42)

class PistachioDataset(VisionDataset):

    def __init__(
            self,
            images
    ) -> None:

        self.images = images
        self.labels = [0]*len(images)

    def __getitem__(self, idx: int):
        """Get the image and label at the given index."""
        # open image using cv2, since opening with Pillow give slightly different results based on Pillow version
        opencv_image = cv2.imread(str(self.images[idx]))
        img = Image.fromarray(cv2.cvtColor(opencv_image, cv2.COLOR_BGR2RGB))
        label = self.labels[idx]

        return img, label

    def __len__(self) -> int:
        """Return the number of images in the dataset."""
        return len(self.images)

def batch_collate(batch):
    imgs, labels = zip(*batch)
    return list(imgs), list(labels)

class PistachioData(ClassificationData):
    def batch_to_labels(self, batch):
        """Convert the batch to a list of labels."""
        return tensor([tensor(x) for x in batch[1]])

    def infer_on_batch(self, batch, model, device):
        """Infer on a batch of images."""
        return None
        
    def batch_to_images(self, batch):
        """Convert the batch to a list of images."""
        return [np.array(x) for x in batch[0]]



root = './datasets/Pistachio_Image_Dataset/all_images'
root = Path(root).absolute()

images_dir = Path(root)

images = list(images_dir.glob('./*.jpg'))
random.shuffle(images)

mid = int(len(images)/2)
train_imgs = images[:mid]
test_imgs = images[mid:]

train_dataloader = DataLoader(dataset=PistachioDataset(images=train_imgs), collate_fn=batch_collate, batch_size=64)
test_dataloader = DataLoader(dataset=PistachioDataset(images=test_imgs), collate_fn=batch_collate, batch_size=64)

train_ds = PistachioData(train_dataloader)
test_ds = PistachioData(test_dataloader)

check = SimilarImageLeakage()
check.run(train_ds, test_ds)
```
